### PR TITLE
Wasm: Implement checked behaviors.

### PR DESCRIPTION
**Pull request by sjrd**
~~Based on #4993 but otherwise ready to review. The first commit belonging to this PR is "Wasm: Implement checked asInstanceOfs.".~~

### CODE REVIEW

**Comment by sjrd** [comment]
This is still based on #4993 but otherwise it's now ready for review.

**Review by tanishiking — COMMENTED** [review]
Mostly nitpicks, looks good from my side

**Comment by sjrd** [comment]
Rebased.

**Comment by sjrd** [comment]
Rebased on top of the non-nullable reference types.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala [hunk 1]
```diff
@@ -518,8 +673,9 @@ private class FunctionEmitter private (
            * point, so we can trap as NPE.
            */
           markPosition(tree)
-          fb += wa.Unreachable
+          genNPE()
         } else {
+          genAsNonNullOrNPEFor(qualifier)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:678]: Why not a check here? IIUC we do not need the as non null in unchecked mode? (it's implicit below).

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala [hunk 2]
```diff
@@ -1920,7 +2079,11 @@ private class FunctionEmitter private (
         fb += wa.GlobalGet(genGlobalID.undef)
 
       case StringType =>
-        fb += wa.RefAsNonNull
+        val sig = watpe.FunctionType(List(watpe.RefType.anyref), List(watpe.RefType.any))
+        fb.block(sig) { nonNullLabel =>
+          fb += wa.BrOnNonNull(nonNullLabel)
+          fb += wa.GlobalGet(genGlobalID.emptyString)
+        }
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:2086]: Hmm... This was a bug, no? If we didn't catch this, doesn't it point to a weakness in the test suite?

Also, should probably be separate?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala [hunk 3]
```diff
@@ -3021,16 +3192,21 @@ private class FunctionEmitter private (
...
+
       case Transients.Cast(expr, tpe) =>
         genCast(expr, tpe, tree.pos)
 
       case value: Transients.SystemArrayCopy =>
         genSystemArrayCopy(tree, value)
 
       case Transients.ObjectClassName(obj) =>
-        genTree(obj, AnyType)
+        genTree(obj, if (obj.tpe.isNullable) AnyType else AnyNotNullType)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:3207]: I feel this appears a bit often. Maybe a helper is appropriate?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala [hunk 4]
```diff
@@ -304,6 +307,155 @@ private class FunctionEmitter private (
   private def receiverStorage: VarStorage.Local =
     _receiverStorage.getOrElse(throw new Error("Cannot access to the receiver in this context."))
 
+  /** Opens a new scope in which NPEs can be thrown by jumping to the NPE label.
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:310]: This is a very nice way of dealing with this.

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala [hunk 5]
```diff
@@ -2600,30 +2606,34 @@ private class FunctionEmitter private (
...
     array.tpe match {
       case ArrayType(arrayTypeRef) =>
-        // Get the underlying array; implicit trap on null
-        fb += wa.StructGet(
-          genTypeID.forArrayClass(arrayTypeRef),
-          genFieldID.objStruct.arrayUnderlying
-        )
+        if (semantics.arrayIndexOutOfBounds == CheckedBehavior.Unchecked) {
+          // Get the underlying array; implicit trap on null
+          fb += wa.StructGet(
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:2840]: Also moved that change to the `nullPointers` commit.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala]: Indeed, a check is enough here.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala (hunk 2 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:2089]: Indeed, that was a bug, although a theoretical one. We have source syntax to emit an `asInstanceOf[string]`, so it is not user-visible.

I did separate it into a commit of its own.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala [hunk 6]
```diff
@@ -383,16 +386,39 @@ object Emitter {
...
+        instantiateClass(ClassCastExceptionClass, StringArgConstructorName)
+      },
+
+      cond(isAnyFatal(asInstanceOfs, arrayIndexOutOfBounds, arrayStores,
+          negativeArraySizes, nullPointers, stringIndexOutOfBounds)) {
+        instantiateClass(UndefinedBehaviorErrorClass,
+            ThrowableArgConsructorName)
+      },
+
+      instantiateClass(ClassClass, AnyArgConstructorName),
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala]: The classes are specified, but not the constructor that we choose to instantiate them. In fact for at least one of them (array index out of bounds, IIRC) we chose a different constructor in Wasm than in JS.

[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala]: Ah, ah :) Fair :)

**Review by gzm0 — APPROVED** [review]

**Pull request merged**